src/battle-scene-interface.ts

```diff
@@ -0,0 +1,176 @@
+	healAnim(pokemon: Pokemon, damage: string | number): void { }
+	hideJoinButtons(): void { }
+	incrementTurn(): void { }
+	log(html: string, preempt?: boolean | undefined): void { }
+	message(message: string, hiddenMessage?: string | undefined): void { }
+	pause(): void { }
+	preemptCatchup(): void { }
+	removeSideCondition(siden: number, id: ID): void { }
+	reset(): void { }
+	resultAnim(pokemon: Pokemon, result: string, type: "bad" | "good" | "neutral" | "par" | "psn" | "frz" | "slp" | "brn"): void { }
+	resume(): void { }
+	runMoveAnim(moveid: ID, participants: Pokemon[]): void { }
+	runOtherAnim(moveid: ID, participants: Pokemon[]): void { }
+	runPrepareAnim(moveid: ID, attacker: Pokemon, defender: Pokemon): void { }
+	runResidualAnim(moveid: ID, pokemon: Pokemon): void { }
+	runStatusAnim(moveid: ID, participants: Pokemon[]): void { }
+	soundStart(): void { }
+	soundStop(): void { }
+	startAnimations(): void { }
+	teamPreview(): void { }
+	teamPreviewEnd(): void { }
+	updateGen(): void { }
+	updateSidebar(side: Side): void { }
+	updateSidebars(): void { }
+	updateStatbars(): void { }
+	updateWeather(instant?: boolean | undefined): void { }
+	upkeepWeather(): void { }
+	wait(time: number): void { }
+	setFrameHTML(html: any): void { }
+	setControlsHTML(html: any): void { }
+	createPokmeonSprite(spriteData: SpriteData | null, pos: InitScenePos, siden: number): IPokemonSprite { return new PokemonSpriteStub(); }
+}
+
+interface IPokemonSprite {
+	updateStatbarIfExists(pokemon: Pokemon, updatePrevhp?: boolean, updateHp?: boolean): void;
+	reset(pokemon: Pokemon): void;
+	updateStatbar(pokemon: Pokemon, updatePrevhp?: boolean, updateHp?: boolean): void;
+	updateHPText(pokemon: Pokemon): void;
+	delay(time: number): this;
+	destroy(): void;
+	anim(end: ScenePos, transition?: string): this;
+	leftof(offset: number): number;
+	behind(offset: number): number;
+	addEffect(id: ID, instant?: boolean): void;
+	removeEffect(id: ID, instant?: boolean): void;
+	clearEffects(): void;
+	animSummon(pokemon: Pokemon, slot: number, instant?: boolean): void;
+	animUnsummon(pokemon: Pokemon, instant?: boolean): void;
+	removeTransform(): void;
+	animDragIn(pokemon: Pokemon, slot: number): void;
+	animDragOut(pokemon: Pokemon): void;
+	animFaint(pokemon: Pokemon): void;
+	animReset(): void;
+	animTransform(pokemon: Pokemon, isCustomAnim?: boolean, isPermanent?: boolean): void;
+	beforeMove(): boolean;
+	afterMove(): boolean;
+
+	top: number;
+	left: number;
+	x: number;
+	y: number;
+	z: number;
+	isMissedPokemon: boolean;
+
+	// The fields below are not used by the headless version
+	sp: SpriteData; // accessed by animation code
+	$statbar: JQuery<HTMLElement> | null; // cannot remove; accessed by BattleScene
+	$el: JQuery<HTMLElement> | null;
+} 
+
+class PokemonSpriteStub implements IPokemonSprite {
+	top: number = NaN;
+	left: number = NaN;
+	x: number = NaN;
+	y: number = NaN;
+	z: number = NaN;
```

Comment:
It seems unnecessary to implement these at all if they're going to just be NaN.

Reply:
The core problem is that the `Pokemon` object holds a reference to its `sprite`, which is used in methods like `BattleScene.abilityActivateAnim` access:

https://github.com/Zarel/Pokemon-Showdown-Client/blob/24ed8577bf51057633be35115bb1d750e1f0ef51/src/battle-animations.ts#L1125-L1133

I went with a solution that I think would minimize the diff size, but I'm open to other solutions. One way to achieve this is to get rid of the `Pokemon.sprite` reference, and have all accesses gated by a `Battle.getSpriteFor(pokemon: Pokemon)` method.

Reply:
This could be implemented with a `Map`, but the client needs to be ES3-compatible.

You would be overriding the relevant `BattleScene` methods, such as `BattleScene#abilityActivateAnim`. I would not implement a sub Sprite at all, just leave `pokemon.sprite = null`. The stub Scene shouldn't be accessing it anyway, after all.

This may involve some complexity with TypeScript – I'm open to either `Pokemon#sprite` being typed as `Sprite` and the null assignment being `pokemon.sprite = null!;` or with `Pokemon#sprite` being typed as `Sprite | null` and the usage being `pokemon.sprite!`.

Reply:
In other words, reading from this stub Sprite at all would be an error. In that case, there's no reason to have it fail buggily rather than have it fail hard.